### ISSUE TITLE
$set->toArray is required for Iterator 

### ISSUE THREAD

**Issue by basz** [php-ds/extension#82]
with the extension v1.1.8 (polypill is fine) I seeing this strange behaviour using php 7.1.3

```
#!/usr/bin/env php
<?php

declare(strict_types=1);

class RolesWhereUpdated
{
    private $additions;

    public static function withChanges(
        \Ds\Set $additions
    ): RolesWhereUpdated {
        $event = new static();

        $event->additions = $additions->toArray();

        return $event;
    }

    public function additions(): \Ds\Set
    {
        return new \Ds\Set($this->additions);
    }
}

$event = RolesWhereUpdated::withChanges(new \Ds\Set(['abc', 'def']));

// not iterated
foreach ($event->additions() as $item => $value) {
    print 'iterated: $event->additions()' . PHP_EOL;
}

// is iterated
foreach ($event->additions()->toArray() as $item => $value) {
    print 'iterated: $event->additions()->toArray()' . PHP_EOL;
}


// but with a plain set
$set = new \Ds\Set(['abc', 'def']);

// this is iterated
foreach ($set as $item => $value) {
    print 'iterated: $set' . PHP_EOL;
}
```

expected output
```
iterated: $event->additions()
iterated: $event->additions()
iterated: $event->additions()->toArray()
iterated: $event->additions()->toArray()
iterated: $set
iterated: $set
```

actual output
```
iterated: $event->additions()->toArray()
iterated: $event->additions()->toArray()
iterated: $set
iterated: $set
```

**Comment by nikic** [comment]
The problem is presumably that https://github.com/php-ds/extension/blob/master/src/php/iterators/php_htable_iterator.c#L160 does not retain a reference to the object, so in this case the object will be destroyed while the iterator is still in use.

**Comment by basz** [comment]
by the way, the same is true for Map

**Comment by rtheunissen** [comment]
@nikic coming back to this, what would be the best way to solve this? I get that we need to hold on to a reference of the object so that it doesn't get collected, so should we add that reference when we create the iterator and release it when we destruct the iterator?

**Comment by rtheunissen** [comment]
@nikic see https://github.com/php-ds/extension/pull/84

**Comment by rtheunissen** [comment]
Can confirm that the fix in #84 produces:
```
iterated: $event->additions()
iterated: $event->additions()
iterated: $event->additions()->toArray()
iterated: $event->additions()->toArray()
iterated: $set
iterated: $set
```

**Comment by rtheunissen** [comment]
This has now been fixed and released as [v1.1.9](https://github.com/php-ds/extension/releases/tag/v1.1.9)
